src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts

```diff
@@ -0,0 +1,95 @@
+      const basePaths = resource.rules
+        .filter(
+          rule =>
+            rule instanceof ruleType &&
+            (/^contained(\[\d+\])?\.resourceType$/.test(getRulePath(rule)) ||
+              /^entry(\[\d+\])?\.resource\.resourceType$/.test(getRulePath(rule)))
+        )
+        .map(rule => getRulePath(rule).replace('.resourceType', ''));
+
+      let generatedIdCount = 0;
+      // For each base path, extract an inline instance
+      basePaths.forEach(basePath => {
+        const rulesToRemove: number[] = [];
+        let id: string;
+        let resourceType: string;
+        let profile: string;
+        const newInstance = new ExportableInstance('');
+
+        // Find all rules on the instance that are children of the base path and should be
+        // added to the inline instance
+        resource.rules.forEach((rule, i) => {
+          if (!(rule instanceof ruleType && getRulePath(rule).startsWith(basePath))) {
+            return;
+          }
+
+          rulesToRemove.push(i);
+          if (rule instanceof ExportableCaretValueRule) {
+            const newRule = new ExportableAssignmentRule(rule.caretPath);
+            newRule.value = rule.value;
+            rule = newRule;
+          }
+          // id and resourceType and meta.profile should be used for keywords, all other rules are added
+          if (rule.path === `${basePath}.id`) {
+            id = rule.value as string;
+          } else if (rule.path === `${basePath}.resourceType`) {
+            resourceType = rule.value as string;
+          } else if (
+            new RegExp(`${escapeRegExp(basePath)}\\.meta\\.profile(\\[0\\])?`).test(rule.path)
+          ) {
+            profile = rule.value as string;
+          } else {
+            newInstance.rules.push(rule);
+          }
+        });
+
+        newInstance.id = id ?? `Inline-Instance-for-${resource.id}-${(generatedIdCount += 1)}`;
+        newInstance.instanceOf = profile ?? resourceType;
+        newInstance.usage = 'Inline';
+        inlineInstances.push(newInstance);
+
+        pullAt(resource.rules, rulesToRemove);
+        if (resource instanceof ExportableInstance) {
+          const inlineInstanceRule = new ExportableAssignmentRule(basePath);
+          inlineInstanceRule.isInstance = true;
+          inlineInstanceRule.value = newInstance.id;
+          resource.rules.push(inlineInstanceRule);
+        } else {
+          const inlineInstanceRule = new ExportableCaretValueRule('');
+          inlineInstanceRule.caretPath = basePath;
+          inlineInstanceRule.isInstance = true;
+          inlineInstanceRule.value = newInstance.id;
+          resource.rules.push(inlineInstanceRule);
+        }
+      });
+    });
+    pkg.instances.push(...inlineInstances);
+  }
+} as OptimizerPlugin;
+
+function getRulePath(rule: ExportableSdRule): string {
+  return rule instanceof ExportableCaretValueRule ? rule.caretPath : rule.path;
```

Comment:
I don't think `caretPath` can be null, but I don't mind being explicit here, even if it's not as clean. I wasn't able to figure out how to get inlining to work, because TypeScript is not sure that the `rule` is a care or assignment, so `rule.caretPath` is not allowed by TS.